tests_tf/test_attacks.py

```diff
@@ -44,6 +57,175 @@ def test_generate_np(self):
 
+class TestFastGradientMethod(CleverHansTest):
+    def setUp(self):
+        super(TestFastGradientMethod, self).setUp()
+        import tensorflow as tf
+
+        # The world's simplest neural network
+        def my_model(x):
+            W1 = tf.constant([[1.5, .3], [-2, 0.3]], dtype=tf.float32)
+            h1 = tf.nn.sigmoid(tf.matmul(x, W1))
+            W2 = tf.constant([[-2.4, 1.2], [0.5, -2.3]], dtype=tf.float32)
+            res = tf.nn.softmax(tf.matmul(x, W2))
+            return res
+
+        self.sess = tf.Session()
+        self.model = my_model
+        self.attack = FastGradientMethod(self.model, sess=self.sess)
+
+    def test_generate_np_gives_adversarial_example(self):
+        x_val = np.random.rand(100, 2)
+        x_val = np.array(x_val, dtype=np.float32)
+
+        for ord in [np.inf, 1, 2]:
+            x_adv = self.attack.generate_np(x_val, eps=.5, ord=ord,
+                                            clip_min=-5, clip_max=5)
+
+            if ord == np.inf:
+                assert np.allclose(np.max(np.abs(x_adv-x_val), axis=1), 0.5)
+            elif ord == 1:
+                assert np.allclose(np.sum(np.abs(x_adv-x_val), axis=1), 0.5)
+            elif ord == 2:
+                assert np.allclose(np.sum(np.square(x_adv-x_val), axis=1)**.5,
+                                   0.5)
+
+            orig_labs = np.argmax(self.sess.run(self.model(x_val)), axis=1)
+            new_labs = np.argmax(self.sess.run(self.model(x_adv)), axis=1)
+
+            assert np.mean(orig_labs==new_labs) < 0.5
+
+    def test_generate_np_can_be_called_with_different_eps(self):
+        x_val = np.random.rand(100, 2)
+        x_val = np.array(x_val, dtype=np.float32)
+
+        for eps in [0.1, 0.2, 0.3, 0.4]:
+            x_adv = self.attack.generate_np(x_val, eps=eps, ord=np.inf,
+                                            clip_min=-5.0, clip_max=5.0)
+
+            assert np.allclose(np.max(np.abs(x_adv-x_val), axis=1), eps)
+
+    def test_generate_np_clip_works_as_expected(self):
+        x_val = np.random.rand(100, 2)
+        x_val = np.array(x_val, dtype=np.float32)
+
+        x_adv = self.attack.generate_np(x_val, eps=0.5, ord=np.inf,
+                                        clip_min=-0.2, clip_max=0.1)
+
+        assert np.isclose(np.min(x_adv), -0.2)
+        assert np.isclose(np.max(x_adv), 0.1)
+
+    def test_generate_np_caches_graph_computation_for_eps_clip_or_xi(self):
+        import tensorflow as tf
+
+        x_val = np.random.rand(1, 2)
+        x_val = np.array(x_val, dtype=np.float32)
+
+        x_adv = self.attack.generate_np(x_val, eps=.3, num_iterations=10,
+                                        clip_max=-5.0, clip_min=-5.0,
+                                        xi=1e-6)
+
+        old_grads = tf.gradients
+
+        def fn(*x, **y):
+            raise RuntimeError()
+        tf.gradients = fn
+
+        x_adv = self.attack.generate_np(x_val, eps=.2, num_iterations=10,
```

Comment:
Should we wrap this call to make sure that the `RuntimError` is not risen using an `assertTrue` on exception not being in the context?